Wobble/Scheduling/TaskHandler.cs

```diff
@@ -0,0 +1,123 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Wobble.Logging;
+
+namespace Wobble.Scheduling
+{
+    /// <summary>
+    ///     This will be used to handle threaded tasks.
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    /// <typeparam name="TResult"></typeparam>
+    public class TaskHandler<T, TResult> : IDisposable
+    {
+        /// <summary>
+        /// This will be used for the task that the TaskHandler will be Handling.
+        /// </summary>
+        private Func<T, TResult> Function { get; }
+
+        /// <summary>
+        ///     Event Invoked when the task gets completed.
+        /// </summary>
+        public event EventHandler<TaskCompleteEventArgs<T, TResult>> OnCompleted;
+
+        /// <summary>
+        ///     Event Invoked when the task gets cancelled.
+        /// </summary>
+        public event EventHandler<TaskCancelledEventArgs<T>> OnCancelled;
+
+        /// <summary>
+        ///     Used to handle Task Cancellation.
+        /// </summary>
+        public CancellationTokenSource Source { get; private set; } = new CancellationTokenSource();
+
+        /// <summary>
+        ///     Determined by if current task is completed.
+        /// </summary>
+        public bool IsCompleted { get; private set; }
+
+        /// <summary>
+        ///     Determined by if the current task is running.
+        /// </summary>
+        public bool IsRunning { get; private set; }
+
+        /// <summary>
+        ///     Determined by if the current task is cancelled.
+        /// </summary>
+        public bool IsCancelled { get; private set; }
+
+        /// <summary>
+        /// </summary>
+        /// <param name="function"></param>
+        public TaskHandler(Func<T, TResult> function) => Function = function;
+
+        /// <summary>
+        ///     This method will cancel the previous task and run a new task.
+        /// </summary>
+        /// <param name="input"></param>
+        /// <param name="delay"></param>
+        public void Run(T input, int delay = 0)
+        {
+            // We want to automatically cancel the previous task and dispose the other
+            // cancellation token source before running a new task.
+            Source.Cancel();
+            Source.Dispose();
+            Source = new CancellationTokenSource();
+            IsCompleted = false;
+            IsCancelled = false;
+            IsRunning = true;
+
+            var token = Source.Token;
+
+            Task.Run(async () =>
+            {
+                try
+                {
+                    await Task.Delay(delay, token);
+
+                    var result = Function(input);
+                    token.ThrowIfCancellationRequested();
+                    IsCompleted = true;
```

Comment:
There's still a race condition between these assignments and the ones in `Run()` (imagine someone calling `Run()` just as these assignments happen). Could you put them in a `lock this {}` block here and above in `Run()`?

Reply:
There shouldn't be a race condition since we're already referencing the old cancellation token. When a new one gets created, there will still be a ref to the old cancellation token and it will cancel and dispose the older one automatically.

Reply:
Sequence of events:
1. Thread A is the currently running task and thread B is calling the `Run()` function.
2. Thread A gets to line 77 and is scheduled out by the OS.
3. Thread B is scheduled and runs from line 61 down. The token is cancelled, but thread A is already past the `ThrowIfCancellationRequested()` line. `IsCompleted` is set to `false`, `IsRunning` is set to `true` and a new task is started by thread B.
4. Thread A is scheduled and runs the lines 77 through 79. `IsCompleted` is set to `true`, `IsRunning` is set to `false`.

The result: `IsCompleted = true`, `IsRunning = false`, but a task is currently running (launched by thread B).

My suggestion: add a `lock (this)` enclosing from `token.ThrowIfCancellationRequested();` to `IsRunning = false;` inclusive and a similar `lock (this)` enclosing `Source.Cancel()` on line 61 so this scenario never happens.